File: frontend/src/types/quest.ts
export type ObjectiveType = "kill" | "fetch" | "talk" | "explore" | "escort" | "craft" | "investigate";
export type QuestType = "main" | "side" | "daily" | "hidden";
export type Difficulty = "easy" | "normal" | "hard" | "legendary";
export type Affinity = "low" | "normal" | "high" | "max";

export interface QuestObjective {
  type: ObjectiveType;
  target: string;
  count: number;
  location?: string;
  hint?: string;
}

export interface QuestBranch {
  condition: string;
  label: string;
  next_quest_seed: string;
  consequence: string;
}

export interface QuestReward {
  gold: number;
  exp: number;
  items: string[];
  affinity_change: number;
  unlock?: string;
}

export interface QuestDialogue {
  on_offer: string;
  on_accept: string;
  on_progress: string;
  on_complete: string;
  on_reject?: string;
}

export interface GeneratedQuest {
  quest_id: string;
  title: string;
  description: string;
  type: QuestType;
  difficulty: Difficulty;
  estimated_time: string;
  prerequisites: string[];
  objectives: QuestObjective[];
  branches?: QuestBranch[];
  rewards: QuestReward;
  dialogue: QuestDialogue;
  lore_connection: string;
  design_notes: string;
}

export interface NpcConfig {
  name: string;
  personality: string;
  occupation: string;
}

export interface WorldConfig {
  genre: string;
  season_or_situation: string;
  special_notes?: string;
}

export interface GameState {
  player_level: number;
  affinity: Affinity;
  quest_type: QuestType;
}

export interface QuestGenerateRequest {
  npc: NpcConfig;
  world: WorldConfig;
  game_state: GameState;
}

export interface Preset {
  id: string;
  name: string;
  npc: NpcConfig;
  world: WorldConfig;
  game_state: GameState;
}

export interface QuestChainResponse {
  chain: GeneratedQuest[];
  chain_narrative: string;
}

export interface QuestContinueRequest {
  quest: GeneratedQuest;
  branch_chosen: string;
  context: QuestGenerateRequest;
}

File: backend/schemas/quest.py
from pydantic import BaseModel
from typing import Literal


class QuestObjective(BaseModel):
    type: Literal["kill", "fetch", "talk", "explore", "escort", "craft", "investigate"]
    target: str
    count: int = 1
    location: str | None = None
    hint: str | None = None


class QuestBranch(BaseModel):
    condition: str
    label: str
    next_quest_seed: str
    consequence: str


class QuestReward(BaseModel):
    gold: int
    exp: int
    items: list[str]
    affinity_change: int
    unlock: str | None = None


class QuestDialogue(BaseModel):
    on_offer: str
    on_accept: str
    on_progress: str
    on_complete: str
    on_reject: str | None = None


class GeneratedQuest(BaseModel):
    quest_id: str
    title: str
    description: str
    type: Literal["main", "side", "daily", "hidden"]
    difficulty: Literal["easy", "normal", "hard", "legendary"]
    estimated_time: str
    prerequisites: list[str]
    objectives: list[QuestObjective]
    branches: list[QuestBranch] | None = None
    rewards: QuestReward
    dialogue: QuestDialogue
    lore_connection: str
    design_notes: str


class QuestChain(BaseModel):
    chain: list[GeneratedQuest]
    chain_narrative: str  # Overall narrative description connecting the quests
